File: Tests/ButterflyNet.Tests/ContextualTests.cs
using NUnit.Framework;

namespace ButterflyNet.Parser.Tests {
	[TestFixture]
	public class ContextualTests : WikiToHtmlTest {

		[Test]
		public void Should_close_list_when_followed_by_block_scope() {
			AssertWithNoRegardForLineBreaks(Parser.ParseAndReturn("* lulz\n! oh hai!"), "<ul><li>lulz</li></ul><h1>oh hai!</h1>");
		}

		[Test]
		public void Should_close_paragraph_before_opening_block_scope() {
			AssertWithNoRegardForLineBreaks(Parser.ParseAndReturn("foo\n*lulz"), "<p>foo</p><ul><li>lulz</li></ul>");
		}

		[Test]
		public void Should_close_list_before_opening_paragraph() {
			const string text = @"*lulz
foo";

			AssertWithNoRegardForLineBreaks(Parser.ParseAndReturn(text), "<ul><li>lulz</li></ul><p>foo</p>");
		}

		[Test]
		public void Should_close_definition_list_before_opening_paragraph() {
			const string text = @";term
:definition
foo";

			AssertWithNoRegardForLineBreaks(Parser.ParseAndReturn(text), "<dl><dt>term</dt><dd>definition</dd></dl><p>foo</p>");
		}

		[Test]
		public void Should_close_paragraph_before_opening_list() {
			const string text = @"foo
* list";

			AssertWithNoRegardForLineBreaks(Parser.ParseAndReturn(text), "<p>foo</p><ul><li>list</li></ul>");
		}

		[Test]
		public void Should_close_paragraph_before_opening_definition_list() {
			const string text = @"foo
;list";

			AssertWithNoRegardForLineBreaks(Parser.ParseAndReturn(text), "<p>foo</p><dl><dt>list</dt></dl>");
		}

		[Test]
		public void Should_handle_contextual_scope_followed_by_paragraph_with_inline_scope() {
			const string text = @"* lol
foo __bar__";

			AssertWithNoRegardForLineBreaks(Parser.ParseAndReturn(text), "<ul><li>lol</li></ul><p>foo <strong>bar</strong></p>");
		}

	}
}

File: src/js/tests/ContextualTests.js
var contextualTests = function() {
	var parser;
	
	return {
		setup: function() {
			parser = createParser();
		},
		
		tearDown: function() {
			parser = null;
		},
		
		test: function Contextual_tests() {
			return [
				function Should_close_list_when_followed_by_block_scope() {
					Assert.that(trimLf(parser.parseAndReturn("* lulz\n! oh hai!")), Is.equalTo("<ul><li>lulz</li></ul><h1>oh hai!</h1>"));
				},
				
				function Should_close_paragraph_before_opening_block_scope() {
					Assert.that(trimLf(parser.parseAndReturn("foo\n*lulz")), Is.equalTo("<p>foo</p><ul><li>lulz</li></ul>"));
				},
				
				function Should_close_list_before_opening_paragraph() {
					Assert.that(trimLf(parser.parseAndReturn("*lulz\nfoo")), Is.equalTo("<ul><li>lulz</li></ul><p>foo</p>"));
				},
				
				function Should_close_definition_list_before_opening_paragraph() {
					Assert.that(trimLf(parser.parseAndReturn(";term\n:definition\nfoo")), Is.equalTo("<dl><dt>term</dt><dd>definition</dd></dl><p>foo</p>"));
				},
				
				function Should_close_paragraph_before_opening_list() {
					Assert.that(trimLf(parser.parseAndReturn("foo\n* list")), Is.equalTo("<p>foo</p><ul><li>list</li></ul>"));
				},
				
				function Should_close_paragraph_before_opening_definition_list() {
					Assert.that(trimLf(parser.parseAndReturn("foo\n;list")), Is.equalTo("<p>foo</p><dl><dt>list</dt></dl>"));
				},
				
				function Should_handle_contextual_scope_followed_by_paragraph_with_inline_scope() {
					Assert.that(trimLf(parser.parseAndReturn("* lol\nfoo __bar__")), Is.equalTo("<ul><li>lol</li></ul><p>foo <strong>bar</strong></p>"));
				}
			];
		}
	};
}();

Jarvis.run(contextualTests);